### PR TITLE
Created versioned ControlPoints classes

### PR DESCRIPTION
**Pull request by jessemapel**
The classes still need to be documented but we need to decide what documentation to put where

### CODE REVIEW

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+   * @return @b QSharedPointer<ControlNetFileProtoV0001_PBControlPoint> A pointer to the internal
+   *                                                                    control point data.
+   */
+  QSharedPointer<ControlNetFileProtoV0001_PBControlPoint> ControlPointV0001::pointData() {
+      return m_pointData;
+  }
+
+
+  /**
+   * This convenience method takes a boolean value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1 protobuf field?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control point that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param point[out] The version 2 protobuf representation of the control point that the value
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1 ?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+    container.deleteKeyword(keyName);
+    value = value.toLower();
+
+    if (value == "true" || value == "yes")
+      (point->*setter)(true);
+  }
+
+
+  /**
+   * This convenience method takes a double value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control point that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param point[out] The version 2 protobuf representation of the control point that the value
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+      return;
+
+    double value = toDouble(container[keyName][0]);
+    container.deleteKeyword(keyName);
+    (point->*setter)(value);
+  }
+
+
+  /**
+   * This convenience method takes a string value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control point that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param point[out] The version 2 protobuf representation of the control point that the value
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+      return;
+
+    QString value = container[keyName][0];
+    container.deleteKeyword(keyName);
+    (point->*setter)(value);
+  }
+
+
+  /**
+   * This convenience method takes a boolean value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control measure that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param[out] measure The version 2 protobuf representation of the control measure that the
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+    container.deleteKeyword(keyName);
+    value = value.toLower();
+
+    if (value == "true" || value == "yes")
+      (measure.*setter)(true);
+  }
+
+
+  /**
+   * This convenience method takes a double value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 10]
```diff
@@ -0,0 +1,638 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control measure that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param[out] measure The version 2 protobuf representation of the control measure that the
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 11]
```diff
@@ -0,0 +1,638 @@
...
+      return;
+
+    double value = toDouble(container[keyName][0]);
+    container.deleteKeyword(keyName);
+    (measure.*setter)(value);
+  }
+
+
+  /**
+   * This convenience method takes a string value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 12]
```diff
@@ -0,0 +1,638 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control measure that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param[out] measure The version 2 protobuf representation of the control measure that the
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: version 1

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+
+    if (!m_pointData->IsInitialized()) {
+      QString msg = "There is missing required information in the control "
+                    "points or measures";
+      throw IException(IException::Io, msg, _FILEINFO_);
+    }
+  }
+
+
+  // TODO finish this once Version 1 is created
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: Maybe make a card for this if you haven't already to add it to our backlog.

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+      if (!m_pointData) {
+        m_pointData.reset(new ControlNetFileProtoV0001_PBControlPoint);
+      }
+
+      return *m_pointData;
+  }
+
+
+  /**
+   * This convenience method takes a boolean value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control point that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param point[out] The version 2 protobuf representation of the control point that the value
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+    container.deleteKeyword(keyName);
+    value = value.toLower();
+
+    if (value == "true" || value == "yes")
+      (point->*setter)(true);
+  }
+
+
+  /**
+   * This convenience method takes a double value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control point that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param point[out] The version 2 protobuf representation of the control point that the value
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 18]
```diff
@@ -0,0 +1,458 @@
...
+      return;
+
+    double value = toDouble(container[keyName][0]);
+    container.deleteKeyword(keyName);
+    (point->*setter)(value);
+  }
+
+
+  /**
+   * This convenience method takes a string value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control point that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param point[out] The version 2 protobuf representation of the control point that the value
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+      return;
+
+    QString value = container[keyName][0];
+    container.deleteKeyword(keyName);
+    (point->*setter)(value);
+  }
+
+
+  /**
+   * This convenience method takes a boolean value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control measure that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param[out] measure The version 2 protobuf representation of the control measure that the
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+    container.deleteKeyword(keyName);
+    value = value.toLower();
+
+    if (value == "true" || value == "yes")
+      (measure.*setter)(true);
+  }
+
+
+  /**
+   * This convenience method takes a double value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control measure that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param[out] measure The version 2 protobuf representation of the control measure that the
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 24]
```diff
@@ -0,0 +1,458 @@
...
+      return;
+
+    double value = toDouble(container[keyName][0]);
+    container.deleteKeyword(keyName);
+    (measure.*setter)(value);
+  }
+
+
+  /**
+   * This convenience method takes a string value from a PvlKeyword and copies it into a version 2
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp [hunk 25]
```diff
@@ -0,0 +1,458 @@
...
+   * protobuf field.
+   *
+   * If the keyword doesn't exist, this does nothing.
+   *
+   * @param container The PvlContainer representation of the control measure that contains the
+   *                  PvlKeyword.
+   * @param keyName The name of the keyword to be copied.
+   * @param[out] measure The version 2 protobuf representation of the control measure that the
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0002.cpp]: version 1?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp [hunk 26]
```diff
@@ -0,0 +1,699 @@
...
+  /**
+   * Create a ControlPointV0003 object from a PvlControlPointV0002 object
+   *
+   * @param oldPoint The PvlControlPointV0002 that will be upgraded to V0003.
+   */
+  ControlPointV0003::ControlPointV0003(ControlPointV0002 &oldPoint)
+   : m_pointData(new ControlPointFileEntryV0002) {
+    QSharedPointer<ControlNetFileProtoV0001_PBControlPoint> oldPointData = oldPoint.pointData();
+
+    // Copy over POD values
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp]: What does `POD` stand for?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp [hunk 27]
```diff
@@ -0,0 +1,699 @@
...
+      m_pointData->add_adjustedcovar( oldPointData->adjustedcovar(5) );
+    }
+
+    // Copy the measures
+    for (int i = 0; i < oldPointData->measures_size(); i++) {
+      ControlPointFileEntryV0002_Measure *newMeasure; = m_pointData->add_measures();
+      ControlNetFileProtoV0001_PBControlPoint_PBControlMeasure oldMeasure;
+      oldMeasure = oldPointData->measures(i);
+
+      // Copy over POD values
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp]: What does `POD` stand for ?

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp [hunk 28]
```diff
@@ -0,0 +1,699 @@
...
+      m_pointData->add_adjustedcovar( oldPointData->adjustedcovar(1) );
+      m_pointData->add_adjustedcovar( oldPointData->adjustedcovar(2) );
+      m_pointData->add_adjustedcovar( oldPointData->adjustedcovar(3) );
+      m_pointData->add_adjustedcovar( oldPointData->adjustedcovar(4) );
+      m_pointData->add_adjustedcovar( oldPointData->adjustedcovar(5) );
+    }
+
+    // Copy the measures
+    for (int i = 0; i < oldPointData->measures_size(); i++) {
+      ControlPointFileEntryV0002_Measure *newMeasure; = m_pointData->add_measures();
```
Comment (kberryUSGS) [inline isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp]: Typo? And I can't tell where the newMeasure actually gets added in this block.

Comment (jessemapel) [reply isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp]: The new measure is added in the m_pointData->add_measures() call. The add_fieldname methods create a new instance, add it to the internal array, and then return a pointer to the new instance. So this line adds a new ControlPointFileEntryV0002_Measure object in the control point's repeated measures field and returns a pointer to it.

See the section titled "The Protocol Buffer API" in the protobuf c++ [documentation/tutorial](https://developers.google.com/protocol-buffers/docs/cpptutorial)

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 29]
```diff
@@ -0,0 +1,638 @@
...
+       // pass to the outer catch
+     }
+     QString msg = "Unable to get target body radii for [" + targetName
+                   + "] when calculating covariance matrix.";
+     throw IException(e, IException::Io, msg, _FILEINFO_);
+    }
+    Distance equatorialRadius(radii["EquatorialRadius"], Distance::Meters);
+    Distance polarRadius(radii["PolarRadius"], Distance::Meters);
+
+    if ( pointObject.hasKeyword("ApostCovarianceMatrix") ) {
```
[the remark this answers is not in the document]
Comment (kberryUSGS) [reply isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: That makes sense! meant to look this up myself before commenting but didn't catch my own note--thanks!

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 30]
```diff
@@ -0,0 +1,638 @@
...
+
+    // Clean up the Pvl control point
+    // Anything that doesn't have a value is removed
+    for (int cpKeyIndex = 0; cpKeyIndex < pointObject.keywords(); cpKeyIndex ++) {
+      if (pointObject[cpKeyIndex][0] == "") {
+        pointObject.deleteKeyword(cpKeyIndex);
+      }
+    }
+
+    // Copy over POD values
```
[the remark this answers is not in the document]
Comment (kberryUSGS) [reply isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: :+1: Thanks!

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp [hunk 31]
```diff
@@ -0,0 +1,638 @@
...
+                                               Distance(sigmaRad, Distance::Meters) );
+      m_pointData->add_aprioricovar( aprioriPoint.GetRectangularMatrix()(0, 0) );
+      m_pointData->add_aprioricovar( aprioriPoint.GetRectangularMatrix()(0, 1) );
+      m_pointData->add_aprioricovar( aprioriPoint.GetRectangularMatrix()(0, 2) );
+      m_pointData->add_aprioricovar( aprioriPoint.GetRectangularMatrix()(1, 1) );
+      m_pointData->add_aprioricovar( aprioriPoint.GetRectangularMatrix()(1, 2) );
+      m_pointData->add_aprioricovar( aprioriPoint.GetRectangularMatrix()(2, 2) );
+    }
+
+    if ( pointObject.hasKeyword("AdjustedSigmaLatitude")
```
[the remark this answers is not in the document]
Comment (kberryUSGS) [reply isis/src/control/objs/ControlNetVersioner/ControlPointV0001.cpp]: Ah, thank you I didn't catch that they were dealing with different matricies.

File: isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp (hunk 28 above)
Comment (kberryUSGS) [reply isis/src/control/objs/ControlNetVersioner/ControlPointV0003.cpp]: :+1: Thanks for the info. I think then all that needs to change is the semicolon should just be removed from the middle of this line?

**Pull request merged**